subsys/logging/log_backend_rtt.c

```diff
@@ -0,0 +1,238 @@
+		*line_pos++ = data;
+	}
+
+	/* not enough space in line buffer, we have to wait for EOL */
+	return 0;
+}
+
+
+static int log_backend_rtt_write_drop(void)
+{
+	*line_pos = '\r';
+
+	if (drop_cnt > 0 && !drop_warn) {
+		memmove(line_buf + DROP_MSG_LEN, line_buf,
+			line_pos - line_buf);
+		memcpy(line_buf, drop_msg, DROP_MSG_LEN);
+		line_pos += DROP_MSG_LEN;
+		drop_warn = 1;
+	}
+
+	if (drop_warn) {
+		int cnt = min(drop_cnt, 99);
+
+		if (cnt < 10) {
+			line_buf[DROP_MSG_LEN - 2] = ' ';
+			line_buf[DROP_MSG_LEN - 3] = (u8_t) ('0' + cnt);
+			line_buf[DROP_MSG_LEN - 4] = ' ';
+		} else {
+			line_buf[DROP_MSG_LEN - 2] = (u8_t) ('0' + cnt % 10);
+			line_buf[DROP_MSG_LEN - 3] = (u8_t) ('0' + cnt / 10);
+			line_buf[DROP_MSG_LEN - 4] = '>';
+		}
+	}
+
+	RTT_LOCK();
+	int ret = SEGGER_RTT_WriteSkipNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER,
+					     line_buf, line_pos - line_buf + 1);
+	RTT_UNLOCK();
+
+	if (!ret) {
+		drop_cnt++;
+		return 0;
+	}
+
+	drop_cnt = 0;
+	drop_warn = 0;
+	return 0;
+}
+
+static int log_backend_rtt_write_block(void)
+{
+	unsigned int ret;
+	*line_pos = '\r';
+
+	RTT_LOCK();
+	ret = SEGGER_RTT_WriteSkipNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER,
+					 line_buf, line_pos - line_buf + 1);
+	RTT_UNLOCK();
+
+	if (ret) {
+		log_backend_rtt_flush();
+		return 0;
+	}
+	return 1;
+}
+
+static int log_backend_rtt_write(void)
+{
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_BLOCK)) {
+		return log_backend_rtt_write_block();
+	} else if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_DROP)) {
+		return log_backend_rtt_write_drop();
+	}
+}
+
+static int log_backend_rtt_panic(u8_t *data, size_t length)
+{
+	unsigned int written;
+
+	/* do not respect mutex, take it over */
+	written = SEGGER_RTT_WriteNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER, data,
+					 length);
+	log_backend_rtt_flush();
+	return written;
+}
+
+
+LOG_OUTPUT_DEFINE(log_output, msg_out, &char_buf, 1);
+
+static void put(const struct log_backend *const backend,
+		struct log_msg *msg)
+{
+	log_msg_get(msg);
+
+	u32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_SHOW_COLOR)) {
+		flags |= LOG_OUTPUT_FLAG_COLORS;
+	}
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP)) {
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+	}
+
+	log_output_msg_process(&log_output, msg, flags);
+
+	log_msg_put(msg);
+
+}
+
+static void log_backend_rtt_cfg(void)
+{
+	SEGGER_RTT_ConfigUpBuffer(CONFIG_LOG_BACKEND_RTT_BUFFER, "Logger",
+				  rtt_buf, sizeof(rtt_buf),
+				  SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+}
+
+static void log_backend_rtt_init(void)
+{
+	SEGGER_RTT_Init();
+
+	if (CONFIG_LOG_BACKEND_RTT_BUFFER > 0) {
+		log_backend_rtt_cfg();
+	}
+
+	panic_mode = 0;
+	line_pos = line_buf;
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	panic_mode = 1;
+}
+
+static void log_backend_rtt_flush(void)
+{
+	while (SEGGER_RTT_HasDataUp(CONFIG_LOG_BACKEND_RTT_BUFFER)) {
+		/* pass */
+	}
+}
+
```

Comment:
I think that log_backend_rtt_flush may be improved like this:

`
while(SEGGER_RTT_HasDataUp(CONFIG_LOG_BACKEND_RTT_BUFFER)) {
     if(!panic_mode) {
          k_yield();
     }
}
`

This ensure that other threads do not starve or depend on round robin scheduling. What do you think?